README.md
QSnake
======

a very simple game written in Qt5 based on these rules:

* Snake grows every 10 steps by one unit
* every step a brick is spawn
* bricks tend to colonize
* every 100 steps a bomb appears
* bombs can be collected and used with the spacebar
* bombs have a radius of 10 units

![](screenshots/2020-02-19_001.png)

# Demo Video
[![QSnake Gameplay Demo](https://img.youtube.com/vi/6lUHXIsVJDg/0.jpg)](https://www.youtube.com/watch?v=6lUHXIsVJDg "QSnake Gameplay Demo")


# Controls
* arrow keys
* spacebar (resets game and triggers bombs)
* P for pausing the game
* ESC quits the game
* 0 resets the high score

# Requirements
Nothing special, a default PC should do it. In the code is still some potential for 
speed improvements. Please notify me, if the game is too laggy on your system. 

# Configuration
Currently all parameters can be configured in the file `QSnake.conf` only, which is by default in 
`~/.config/KarlZeilhofer/` on Linux or `%APPDATA%\KarlZeilhofer` on Windows. 

A typical config could be:
```
[General]
BombInterval=100
ExplosionRadius=10
BoxSize=8
BrickAttraction=15
BrickSize=8
GrowInterval=10
HeadClearance=7
InitialLength=10
SizeX=100
SizeY=60
delayEasy=250
delayExtreme=50
delayHard=100
delayMedium=150
difficulty=1
highScore=143
```

Delays are in milli seconds.  
Dimensions are in multiple of `BoxSize` and `BoxSize` itself is in pixels.  
`difficulty` could be 0 to 3 for easy to extreme.  
`BrickAttraction` is the inverse of the probability for a brick middle 
in nowhere without any neighbors. 


# Building on Linux

```
qmake
make
```

# License
public domain




CHANGELOG.md
Changelog of QSnake
===========================

# v0.0.5
* add help texts when died and paused
* bugfix of help texts
* immediate restart on key R

# v0.0.4
* improve resume on keypress
* respect grow and bomb interval
* Reset game now with key R
* add option for self collision
* increase snake body with overlapping circles
* fadeout of explosion

# v0.0.3
* some improvements of the code and usability
* central color definitions
* all values are configurable via .conf file
* store/restore difficulty
* improve colors
* pause and score overlay
* set windowTitle with version
* bugfix (bomb on brick), add difficulty menu, add score overlay
* add screenshot
* first version


scene.cpp
#include "scene.hh"

#include <qgraphicsitem.h>

Scene::Scene()
{
	reset();
	setBackgroundBrush(QColor(QColor(50,50,50)));
}

void Scene::moveSnake()
{
	if(dead){
		return;
	}
	
	static int cnt=0;
	
	if(steerQueue.empty() == false){
		currentHeadDirection = steerQueue.first();
		steerQueue.removeFirst();
	}
	
	QPoint newHead;
	switch(currentHeadDirection){
	case Up:
		newHead.setX(snake.first().x());
		newHead.setY(snake.first().y()-1);
		break;
	case Down:
		newHead.setX(snake.first().x());
		newHead.setY(snake.first().y()+1);
		break;
	case Left:
		newHead.setX(snake.first().x()-1);
		newHead.setY(snake.first().y());
		break;
	case Right:
		newHead.setX(snake.first().x()+1);
		newHead.setY(snake.first().y());
		break;
	}
	
	
	// check for collision
	if(snake.contains(newHead) ||
			bricks.contains(newHead)){
		die();
		return;
	}
	
	snake.insert(0, newHead);
	if(cnt%10 != 0){
		snake.removeLast();
	}
	addBrick();
	if(cnt%100 == 0){
		addBomb();
	}
	
	
	//   with bombs
	if(bombs.contains(newHead)){
		bombsCounter++;
		bombs.removeOne(newHead);
	}
	
	
	paint();
	
	cnt++;
}

void Scene::changeDirection(Scene::Direction dir)
{
	Direction lastDir;
	if(steerQueue.length() >= 1){
		lastDir = steerQueue.last();
	}else{
		lastDir = currentHeadDirection;
	}
	
	if((dir == Left && lastDir != Right) ||
			(dir == Right && lastDir != Left) ||
			(dir == Up && lastDir != Down) ||
			(dir == Down && lastDir != Up)
			){
		if(dir != lastDir){
			steerQueue.append(dir);
		}
	}
}

void Scene::reset()
{
	dead = false;
	snake.clear();
	bombsCounter = 0;
	bombs.clear();
	lastBomb = QPoint();
	
	for(int i=0; i<InitialLength; i++){
		snake.insert(0,QPoint(i+SizeX/2,SizeY/2));
	}
	
	steerQueue.clear();
	steerQueue.append(Right);
	bricks.clear();
	for(int x=0; x<SizeX; x++){
		bricks.insert(0,QPoint(x,0));
		bricks.insert(0,QPoint(x,SizeY));
	}
	for(int y=0; y<SizeY; y++){
		bricks.insert(0,QPoint(0,y));
		bricks.insert(0,QPoint(SizeX,y));
	}
}

void Scene::addBrick()
{
	int neighbourSearchTrials = BrickAttraction; // countdown for searching neighbours, if zero, we will plant a hermit in the middle of nowhere
	
	int retries = 1000;
	while(retries--){
		QPoint p((rand()%(SizeX-1))+1, 
				 (rand()%(SizeY-1))+1);
		
		// check for enough distance to head, to avoid unfair obstacles
		if(abs(p.x()-snake.first().x()) <= HeadClearance &&
				abs(p.y()-snake.first().y()) <= HeadClearance){
			continue;
		}
		
		// check random position against snake, bricks and bombs
		if(snake.contains(p) ||
				bricks.contains(p) ||
				bombs.contains(p)){
			continue;
		}
		
		// search for neigbours
		int found=0;
		const int r=1; // search radius
		
		for(int i=-r; i<=r; i++){
			for(int j=-r; j<=r; j++){
				if(i!=0 && j!=0){ // skip center and diagonals
					if(bricks.contains(QPoint(p.x()+i,p.y()+j))){
						found++;
					}
				}
			}
		}
		
		if(found >= 1 || neighbourSearchTrials<=0 ){
			bricks.append(p);
			break;
		}
		neighbourSearchTrials--;
	}
}

void Scene::addBomb()
{
	int retries = 1000;
	while(retries--){
		QPoint p((rand()%(SizeX-1))+1, 
				 (rand()%(SizeY-1))+1);
		
		// check random position against snake, bricks and bombs
		if(snake.contains(p) == false &&
				bricks.contains(p) == false &&
				bombs.contains(p) == false){
			bombs.append(p);
			break;
		}
	}
}

void Scene::triggerBomb()
{
	int r=BombRadius;
	
	if(bombsCounter > 0){
		for(int i=0; i<bricks.length(); i++){
			QPoint p = bricks[i];
			if(p.x() == 0 || p.x() == SizeX || p.y()==0 || p.y()==SizeY){
				continue;
			}
			
			int dx = snake.first().x() - p.x();
			int dy = snake.first().y() - p.y();
			if(dx*dx + dy*dy <= r*r){
				bricks.removeAt(i);
				i--;
			}
		}		
		bombsCounter--;
		lastBomb = snake.first();
	}
}

void Scene::paint()
{
	clear();
	
	// Snake
	for(int i=0; i<snake.length(); i++){
		QPoint p = snake[i];
		QBrush brush(QColor(200, 255, 200));
		if(i==0){
			brush = QBrush(Qt::red);
		}else if(i <= bombsCounter){
			// stripes of the snake
			brush = QBrush(QColor(150,150,255));
		}else if(i%5 == 0){
			// stripes of the snake
			brush = QBrush(QColor(255,220,220));
		}
		auto item = new QGraphicsEllipseItem(BoxSize*p.x(), BoxSize*p.y(), BoxSize, BoxSize);
		item->setBrush(brush);
		addItem(item);
	}
	
	// Bricks
	for(int i=0; i<bricks.length(); i++){
		QPoint p = bricks[i];
		QBrush brush = QBrush(Qt::gray);
		
		auto item = new QGraphicsRectItem(BoxSize*p.x(), BoxSize*p.y(), BoxSize, BoxSize);
		item->setBrush(brush);
		addItem(item);
	}
	
	// Bombs
	for(int i=0; i<bombs.length(); i++){
		QPoint p = bombs[i];
		QBrush brush = QBrush(QColor(150,150,255));
		
		auto item = new QGraphicsEllipseItem(BoxSize*p.x(), BoxSize*p.y(), BoxSize, BoxSize);
		item->setBrush(brush);
		addItem(item);
	}
	
	// explosion radius
	if(lastBomb.isNull() == false){
		QBrush brush(QBrush(QColor(150,150,255)));
		int d = BoxSize*BombRadius;
		auto item = new QGraphicsEllipseItem(BoxSize*lastBomb.x()-d, BoxSize*lastBomb.y()-d, 
											 BoxSize*2*BombRadius, BoxSize*2*BombRadius);
		item->setBrush(brush);
		addItem(item);
		lastBomb = QPoint();
	}
	
	// Print Score
	if(dead){
		QBrush brush(Qt::white);
		auto item = new QGraphicsSimpleTextItem(QString("Score %1").arg(length()));
		item->setBrush(brush);
		item->setFont(QFont("DejaVu Sans Mono, Bold", 64, 5));
		QRectF bR = item->sceneBoundingRect();
		item->setPos(QPoint(BoxSize*SizeX/2 - int(bR.width()/2), 
							BoxSize*SizeY/2 - int(bR.height()/2)));
		addItem(item);
	}
}

int Scene::length()
{
	return snake.length();
}

void Scene::die()
{
	dead = true;
	paint();
}

bool Scene::isDead()
{
	return dead;
}


qsnake.cpp
#include "qsnake.hh"
#include "ui_qsnake.h"

#include "scene.hh"

#include <QGraphicsRectItem>
#include <QKeyEvent>
#include <QDebug>
#include <QSettings>
#include <QActionGroup>

QSnake::QSnake(QWidget *parent)
	: QMainWindow(parent)
	, ui(new Ui::QSnake)
{
	ui->setupUi(this);
	scene = new Scene();
	ui->view->setScene(scene);
	
	srand(uint32_t(time(nullptr)));
	
	scene->addItem(new QGraphicsRectItem(50, 50, 60, 40, nullptr));
	
	QActionGroup* ag_difficulty =  new QActionGroup(this);
	ag_difficulty->addAction(ui->actionEasy);
	ag_difficulty->addAction(ui->actionMedium);
	ag_difficulty->addAction(ui->actionHard);
	ag_difficulty->addAction(ui->actionExtreme);
	
	timer = new QTimer(this);
	connect(timer, SIGNAL(timeout()), this, SLOT(timeStep()));
	timer->start(250);
	
	// catch all events by our own eventFilter()
	qApp->installEventFilter(this);
	
	score = new QLabel("Score: 0", this);
	ui->statusbar->addWidget(score);
	
	setWindowTitle(QApplication::applicationName() + " V" + QApplication::applicationVersion());
	showMaximized();
}

QSnake::~QSnake()
{
	delete ui;
}

void QSnake::timeStep()
{
	if(!paused){
		scene->moveSnake();	
		setScore();
	}
}

bool QSnake::eventFilter(QObject *watched, QEvent *event)
{
	
	if(event->type() == QEvent::KeyPress){
		qDebug() << "key event";
		QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
		
		if( keyEvent->key() == Qt::Key_Up )
		{
			scene->changeDirection(Scene::Up);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_Down )
		{
			scene->changeDirection(Scene::Down);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_Left )
		{
			scene->changeDirection(Scene::Left);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_Right )
		{
			scene->changeDirection(Scene::Right);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_P )
		{
			pause();
			return true;
		}else if( keyEvent->key() == Qt::Key_Space )
		{
			if(scene->isDead()){
				qDebug() << "reset()";
				scene->reset();
			}else {
				scene->triggerBomb();
			}

			return true;
		}else if( keyEvent->key() == Qt::Key_0 )
		{
			QSettings set;
			set.setValue("highScore", 0);
			return true;
		}
	}
	return qApp->eventFilter( watched, event );
}

void QSnake::pause()
{
	paused = !paused;
}

void QSnake::setScore()
{
	QSettings set;
	int hs = set.value("highScore").toInt();
	
	int l = scene->length();
	score->setText(QString("Score: %1    High Score: %2").arg(l).arg(hs));
	
	if(l > hs){
		set.setValue("highScore", l);
	}
}



void QSnake::on_actionEasy_triggered()
{
    delay_ms = 250;
	timer->setInterval(delay_ms);
}

void QSnake::on_actionMedium_triggered()
{
    delay_ms = 150; 
	timer->setInterval(delay_ms);
}

void QSnake::on_actionHard_triggered()
{
    delay_ms = 100; 
	timer->setInterval(delay_ms);
}

void QSnake::on_actionExtreme_triggered()
{
    delay_ms = 50; 
	timer->setInterval(delay_ms);
}


mainwindow.cpp
#include "mainwindow.hh"
#include "ui_qsnake.h"

#include "game.hh"

#include <QGraphicsRectItem>
#include <QKeyEvent>
#include <QDebug>
#include <QSettings>
#include <QActionGroup>
#include <QtOpenGL>

MainWindow::MainWindow(QWidget *parent)
	: QMainWindow(parent)
	, ui(new Ui::QSnake)
{
	ui->setupUi(this);
	game = new Game();
	ui->view->setScene(game);
	ui->view->setRenderHint(QPainter::Antialiasing);
	//ui->view->setViewport(new QGLWidget(this));
	
	srand(uint32_t(time(nullptr)));
	
	QActionGroup* ag_difficulty =  new QActionGroup(this);
	ag_difficulty->addAction(ui->actionEasy);
	ag_difficulty->addAction(ui->actionMedium);
	ag_difficulty->addAction(ui->actionHard);
	ag_difficulty->addAction(ui->actionExtreme);
	
	QSettings set;
	
	Difficulty d = Difficulty(set.value("difficulty", 2).toInt());
	
	timer = new QTimer(this);
	connect(timer, SIGNAL(timeout()), this, SLOT(timeStep()));
	setDifficulty(d);

	ui->actionselfcollision->setChecked(set.value("selfCollision", 1).toInt());
	game->setSelfCollision(ui->actionselfcollision->isChecked());
	
	// catch all events by our own eventFilter()
	qApp->installEventFilter(this);
	
	score = new QLabel("Score: 0", this);
	ui->statusbar->addWidget(score);
	
	setWindowTitle(QApplication::applicationName() + " V" + QApplication::applicationVersion());
	showMaximized();
	
	timer->start();
}

MainWindow::~MainWindow()
{
	delete ui;
}

void MainWindow::timeStep()
{
	game->timeStep(timer->interval()*0.001);	
	setScore();
}

bool MainWindow::eventFilter(QObject *watched, QEvent *event)
{
	if(event->type() == QEvent::KeyPress){
		QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
		if(keyEvent->isAutoRepeat()){
			return true; // consume event
		}
		if( keyEvent->key() == Qt::Key_Up )
		{
			game->controlInput(Game::Boost);
			return true; // consume event
		}else 	if( keyEvent->key() == Qt::Key_Down )
		{
			return true; // consume event
		}else 	if( keyEvent->key() == Qt::Key_Left )
		{
			game->controlInput(Game::Left);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_Right )
		{
			game->controlInput(Game::Right);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_P )
		{
			game->togglePause();
			return true;
		}else if( keyEvent->key() == Qt::Key_R )
		{
			game->restart();
			game->setPause(false);
			return true;
		}else if( keyEvent->key() == Qt::Key_Space )
		{
			game->triggerBomb();
			return true;
		}else if( keyEvent->key() == Qt::Key_0 )
		{
			QSettings set;
			set.setValue("highScore", 0);
			return true;
		}else if( keyEvent->key() == Qt::Key_Escape )
		{
			QApplication::exit();
			return true;
		}
	}
	
	if(event->type() == QEvent::KeyRelease){
		QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
		
		if(keyEvent->isAutoRepeat()){
			return false;
		}
		
		if( keyEvent->key() == Qt::Key_Up )
		{
			game->controlInput(Game::NormalSpeed);
			return true; // consume event
		}else if( keyEvent->key() == Qt::Key_Left )
		{
			game->controlInput(Game::Straight);
			return true;
		}else 	if( keyEvent->key() == Qt::Key_Right )
		{
			game->controlInput(Game::Straight);
			return true;
		}
	}
	
	return qApp->eventFilter( watched, event );
}

void MainWindow::setScore()
{
	QSettings set;
	int hs = set.value("highScore").toInt();
	
	int l = game->score();
	score->setText(QString("Score: %1    High Score: %2").arg(l).arg(hs));
	
	if(l > hs){
		set.setValue("highScore", l);
	}
}

void MainWindow::setDifficulty(MainWindow::Difficulty d)
{
	QSettings set;
	
	switch (d) {
	case Easy:
		delay_ms = set.value("delayEasy", 250).toInt();
		set.setValue("delayEasy", delay_ms);
		ui->actionEasy->setChecked(true);
		break;
	case Medium:
		delay_ms = set.value("delayMedium", 150).toInt();
		set.setValue("delayMedium", delay_ms);
		ui->actionMedium->setChecked(true);
		break;
	case Hard:
		delay_ms = set.value("delayHard", 100).toInt();
		set.setValue("delayHard", delay_ms);
		ui->actionHard->setChecked(true);
		break;
	case Extreme:
		delay_ms = set.value("delayExtreme", 50).toInt();
		set.setValue("delayExtreme", delay_ms);
		ui->actionExtreme->setChecked(true);
		break;
	}
	timer->setInterval(33);
	
	set.setValue("difficulty", int(d));
}



void MainWindow::on_actionEasy_triggered()
{
	setDifficulty(Easy);
}

void MainWindow::on_actionMedium_triggered()
{
	setDifficulty(Medium);
}

void MainWindow::on_actionHard_triggered()
{
	setDifficulty(Hard);
}

void MainWindow::on_actionExtreme_triggered()
{
	setDifficulty(Extreme);
}

void MainWindow::on_actionRestore_Defaults_triggered()
{
    game->restoreDefaults();
}

void MainWindow::on_actionselfcollision_triggered(bool checked)
{
    game->setSelfCollision(checked);
	QSettings set;
	set.setValue("selfCollision", checked?1:0);
}


game.cpp
#include "game.hh"

#include <qgraphicsitem.h>
#include <QSettings>
#include <math.h>
#include <QDebug>

Game::Game()
{
	BombColor = QColor(Qt::yellow);
	SnakeColor1 = QColor(100, 255, 100);
	SnakeColor2 = QColor(255,100,100);
	SnakeHeadColor = Qt::red;
	BackgroundColor = QColor(50,50,50);
	BrickColor = Qt::gray;
	TextColor = Qt::white;
	
	setBackgroundBrush(BackgroundColor);
	
	setDefaults();
	QSettings set;
	
	valueNames << "ExplosionRadius";
	valueNames << "SizeX";
	valueNames << "SizeY";
	valueNames << "BrickSize";
	valueNames << "BrickAttraction";
	valueNames << "HeadClearance";
	valueNames << "InitialLength";
	valueNames << "GrowInterval";
	valueNames << "BombInterval";
	
	values.append(&ExplosionRadius);
	values.append(&SizeX);
	values.append(&SizeY);
	values.append(&BrickSize);
	values.append(&BrickAttraction);
	values.append(&HeadClearance);
	values.append(&InitialLength);
	values.append(&GrowInterval);
	values.append(&BombInterval);
	values.append(&BrickInterval);
	
	for(int i=0; i<valueNames.length(); i++){
		if(set.contains(valueNames[i])){
			*(values[i]) = set.value(valueNames[i]).toDouble();
		}else{
			set.setValue(valueNames[i], *(values[i]));
		}
	}
	
	// TODO remove temporary:
	restoreDefaults();
	
	restart();
}

void Game::timeStep(qreal dt)
{
	if(dead || paused){
		paint();
		return;
	}
	
	// process steering
	qreal roundsPerSecond = 0.6; 
	if(currentSteering == Left){
		currentHeadDirection += 360*dt*roundsPerSecond;
		if(currentHeadDirection >= 360){
			currentHeadDirection -= 360;
		}
	}else if(currentSteering == Right){
		currentHeadDirection -= 360*dt*roundsPerSecond;
		if(currentHeadDirection<0){
			currentHeadDirection+=360;
		}
	}
	
	static int cnt=0;
	
	QPointF newHead;
	qreal s = speed * (boost?2:1);
	qreal dx = s*dt*cos(currentHeadDirection*M_PI/180.0);
	qreal dy = -s*dt*sin(currentHeadDirection*M_PI/180.0);
	QPointF d(dx,dy);
	
	// TODO: good creep simulation
//	if((cnt/snake.length())%2 == 0){
//		d *= 0.5; // creep simulation
//	}
	
	newHead = snake.first() + d;
	
	
	// check for self collision
	if(selfCollision){
		QList<QPointF> subList = snake.mid(20);
		if(isColliding(subList, newHead, SnakeSize) >= 0){
			die();
			return;
		}
	}
	// collision with bricks:
	for(int i=0; i<bricks.length(); i++){
		if(isColliding(bricks, newHead, (SnakeSize/2 + BrickSize/2)*0.8) >= 0){
			die();
			return;
		}
	}

	
	
	static qreal brickTimer=0;
	brickTimer += dt;
	if(brickTimer >= BrickInterval){
		brickTimer = 0;
		addBrick();
	}

	snake.insert(0, newHead);	
	static qreal growTimer=0;
	growTimer += dt;
	if(growTimer >= GrowInterval){
		growTimer = 0;
	}else{
		snake.removeLast();
	}
	
	static qreal bombTimer=0;
	bombTimer += dt;
	if(bombTimer >= BombInterval){
		bombTimer = 0;
		addBomb();
	}
	

	// collision with bombs
	int i = isColliding(bombs, newHead, BrickSize/2+SnakeSize/2);
	if(i >= 0){
		bombsCounter++;
		bombs.removeAt(i);
	}

	
	paint();
	
	cnt++;
}

void Game::controlInput(Game::Controlls dir)
{
	if(paused){
		togglePause();
	}
	
	switch(dir){
	case Boost: boost = true; break;
	case NormalSpeed: boost = false; break;
	case Left: 
	case Right: 
	case Straight: currentSteering = dir;
	}
	
}

void Game::restart()
{
	dead = false;
	snake.clear();
	bombsCounter = 10;
	bombFadoutCounter=BombFadeout;
	bombs.clear();
	lastBomb = QPointF();
	currentHeadDirection = 0;
	
	for(int i=0; i<InitialLength; i++){
		snake.insert(0,QPointF(SnakeSize*(i+5),BrickSize*SizeY/2));
	}
	
	bricks.clear();
	for(int x=0; x<=SizeX; x++){
		bricks.insert(0,QPointF(x, 0)*BrickSize);
		bricks.insert(0,QPointF(x, SizeY)*BrickSize);
	}
	for(int y=1; y<SizeY; y++){
		bricks.insert(0,QPointF(0,y)*BrickSize);
		bricks.insert(0,QPointF(SizeX,y)*BrickSize);
	}
	paused = true;
}

void Game::addBrick()
{
	int neighbourSearchTrials = int(BrickAttraction); // countdown for searching neighbours, if zero, we will plant a hermit in the middle of nowhere
	
	int retries = 1000;
	while(retries--){
		QPointF p((rand()%(int(SizeX*BrickSize-BrickSize)))+BrickSize, 
				 (rand()%(int(SizeY*BrickSize-BrickSize)))+BrickSize);
		
		// check for enough distance to head, to avoid unfair obstacles
		QPointF d = p-snake.first();
		if(d.x()*d.x() + d.y()*d.y() < (HeadClearance*BrickSize)*(HeadClearance*BrickSize)){
			continue;
		}
		
		// check random position against snake, bricks and bombs
		if(isColliding(snake, p, (SnakeSize/2+BrickSize/2)*1.2) >= 0 ||
				isColliding(bricks, p, BrickSize*1.2) >= 0 ||
				isColliding(bombs, p, BrickSize*1.2) >= 0){
			continue;
		}
		
		// search for neigbours
		int found=0;
		qreal r=BrickSize*2; // search radius
		
		if(isColliding(bricks, p, r) >= 0){
			found = 1;
		}
		
		if(found >= 1 || neighbourSearchTrials<=0 ){
			bricks.append(p);
			break;
		}
		neighbourSearchTrials--;
	}
}

void Game::addBomb()
{
	int retries = 1000;
	while(retries--){
		QPointF p((rand()%(int(SizeX*BrickSize-BrickSize)))+BrickSize, 
				 (rand()%(int(SizeY*BrickSize-BrickSize)))+BrickSize);
		
		// check random position against snake, bricks and bombs
		if(isColliding(snake, p, (SnakeSize/2+BrickSize/2)*1.2) >= 0 ||
				isColliding(bricks, p, BrickSize*1.2) >= 0 ||
				isColliding(bombs, p, BrickSize*1.2) >= 0){
			continue;
		}

		bombs.append(p);
		break;
	}
}

void Game::triggerBomb()
{
	if(dead || paused)
		return;
	
	qreal r=ExplosionRadius;
	
	if(bombsCounter > 0){
		for(int i=0; i<bricks.length(); i++){
			QPointF p = bricks[i];
			if(round(p.x()) == 0 || round(p.x()) == SizeX*BrickSize || 
					round(p.y())==0 || round(p.y())==SizeY*BrickSize){
				continue;
			}
			

			if(isColliding(snake.first(), p, ExplosionRadius*BrickSize)){
				bricks.removeAt(i);
				i--;
			}
		}		
		bombsCounter--;
		lastBomb = snake.first();
		bombFadoutCounter = BombFadeout;
	}
	paint(); // immediately paint the bomb explosion
}

void Game::paint()
{
	clear();
	
	// Bricks
	for(int i=0; i<bricks.length(); i++){
		QPointF p = bricks[i];
		QBrush brush = QBrush(BrickColor);
		
		auto item = new QGraphicsEllipseItem(p.x()-0.5*BrickSize, p.y()-0.5*BrickSize, 
										  BrickSize, BrickSize);
		item->setBrush(brush);
		addItem(item);
	}
	
	// Bombs
	for(int i=0; i<bombs.length(); i++){
		QPointF p = bombs[i];
		QBrush brush = QBrush(BombColor);
		
		auto item = new QGraphicsEllipseItem(p.x()-0.5*BrickSize, p.y()-0.5*BrickSize, 
											 BrickSize, BrickSize);
		item->setBrush(brush);
		addItem(item);
	}
	
	// TODO IMPORTANT: improve painting of snake!
	// starts to lag at about length 500
	// Snake
	for(int i=snake.length()-1; i>=0; i--){
		QPointF p = snake[i];
		QBrush brush(SnakeColor1);
		
		if(i==0){
			brush = QBrush(SnakeHeadColor);
		}else if(i <= bombsCounter){
			// collected bombs
			brush = QBrush(BombColor);
		}else if((i/10)%2 == 0){
			// stripes of the snake
			brush = QBrush(SnakeColor2);
		}
		auto item = new QGraphicsEllipseItem(p.x()-0.65*SnakeSize, p.y()-0.65*SnakeSize, 
											 SnakeSize*1.3, SnakeSize*1.3);
		item->setBrush(brush);
		addItem(item);
	}
	
	// explosion radius
	if(lastBomb.isNull() == false){
		QColor c = BombColor;
		c.setAlpha(128*bombFadoutCounter/BombFadeout);
		QBrush brush(c);
		
		qreal d = BrickSize*ExplosionRadius;
		auto item = new QGraphicsEllipseItem(lastBomb.x()-d, lastBomb.y()-d, 
											 2*d, 2*d);
		item->setBrush(brush);
		addItem(item);
		bombFadoutCounter--;
		
		if(bombFadoutCounter == 0){
			lastBomb = QPointF();
			bombFadoutCounter = BombFadeout;
		}
	}
	
	// dark overlay
	if(dead || paused){
		QBrush brush(QColor(0,0,0,128)); // transparent black as text background color
		auto rect = new QGraphicsRectItem(-BrickSize*0.5,-BrickSize*0.5,
										  BrickSize*(SizeX+1), BrickSize*(SizeY+1));
		rect->setBrush(brush);
		addItem(rect);
		
	}
	// Print Score
	if(dead){
		{
			auto brush = QBrush(TextColor);
			auto item = new QGraphicsSimpleTextItem(QString("SCORE %1").arg(score()));
			item->setBrush(brush);
			item->setFont(QFont("DejaVu Sans Mono, Bold", 64, 5));
			QRectF bR = item->sceneBoundingRect();
			item->setPos(QPointF(BrickSize*SizeX/2 - int(bR.width()/2), 
								BrickSize*SizeY/2 - int(bR.height()/2)));
			addItem(item);
		}
		{
			auto brush = QBrush(TextColor);
			auto item = new QGraphicsSimpleTextItem(QString("Press R to restart\n    ESC to quit"));
			item->setBrush(brush);
			item->setFont(QFont("DejaVu Sans Mono, Bold", 16, 5));
			QRectF bR = item->sceneBoundingRect();
			item->setPos(QPointF(BrickSize*SizeX/2 - int(bR.width()/2), BrickSize*(SizeY-1) - bR.height()));
			addItem(item);
		}
	}
	
	// Print Pause
	if(paused && !dead){
		auto brush = QBrush(TextColor);
		auto item = new QGraphicsSimpleTextItem(QString("PAUSED"));
		item->setBrush(brush);
		item->setFont(QFont("DejaVu Sans Mono, Bold", 64, 5));
		QRectF bR = item->sceneBoundingRect();
		item->setPos(QPointF(BrickSize*SizeX/2 - bR.width()/2, 
							BrickSize*SizeY/2 - bR.height()/2));
		addItem(item);
		
		{
			auto brush = QBrush(TextColor);
			auto item = new QGraphicsSimpleTextItem(QString("Press any cursor key to start\n        ESC to quit"));
			item->setBrush(brush);
			item->setFont(QFont("DejaVu Sans Mono, Bold", 16, 5));
			QRectF bR = item->sceneBoundingRect();
			item->setPos(QPointF(BrickSize*SizeX/2 - bR.width()/2, 
								 BrickSize*(SizeY-1) - bR.height()));
			addItem(item);
		}
	}
}

int Game::score()
{
	return snake.length();
}

void Game::die()
{
	dead = true;
	paint();
}

void Game::togglePause()
{
	if(dead){
		paused = false;
	}else{
		paused = !paused;
	}
	paint();
}

void Game::setPause(bool flag)
{
	paused = flag;
}

bool Game::isDead()
{
	return dead;
}

void Game::restoreDefaults()
{
	QSettings set;
	die();
	
	setDefaults();
	
	for(int i=0; i<valueNames.length(); i++){
		set.setValue(valueNames[i], *(values[i]));
	}
	
	restart();
}

void Game::setSelfCollision(bool flag)
{
	selfCollision = flag;
}

void Game::setDefaults()
{
	ExplosionRadius=7; // brick units
	SizeX = 100; // brick units
	SizeY = 60; // brick units
	BrickSize = 16; // pixels
	SnakeSize = 16; // pixels
	BrickAttraction = 15; // number of retries for finding neighbours, before settling into nowhere
	HeadClearance = 8; // zone around head, forbidden for new bricks
	InitialLength = 70;	
	GrowInterval = 0.2; // grow snake every n seconds
	BombInterval = 10.0; // spawn a bomb on the field every n seconds
	BrickInterval = 0.40; // a brick every second
}

bool Game::isColliding(QPointF &p1, QPointF &p2, qreal radius)
{
	QPointF d = p1-p2;
	if(d.x()*d.x() + d.y()*d.y() < radius*radius){
		return true;
	}else{
		return false;
	}
}

int Game::isColliding(QList<QPointF> &list, QPointF &p2, qreal radius)
{
	for(int i=0; i<list.length(); i++){
		if(isColliding(list[i], p2, radius)){
			return i;
		}
	}
	return -1;
}
